File: merge_available_branches.py
#!/usr/bin/env python3
"""
Efficient script to merge available cursor branches into main
Handles large numbers of branches systematically
"""

import subprocess
import sys
import time
from datetime import datetime

def run_command(cmd, timeout=30):
    """Run a shell command and return the result"""
    try:
        result = subprocess.run(
            cmd,
            shell=True,
            capture_output=True,
            text=True,
            timeout=timeout
        )
        return result.returncode == 0, result.stdout, result.stderr
    except subprocess.TimeoutExpired:
        return False, "", "Command timed out"
    except Exception as e:
        return False, "", str(e)

def get_available_branches():
    """Get list of available cursor branches"""
    success, output, error = run_command("git branch -r | grep 'cursor/fix-errors-and-merge-to-main' | head -100")
    if not success:
        print(f"Error getting branches: {error}")
        return []
    
    branches = []
    for line in output.strip().split('\n'):
        if line.strip():
            branch_name = line.strip().replace('origin/', '')
            branches.append(branch_name)
    
    return branches

def merge_branch(branch_name):
    """Attempt to merge a single branch"""
    print(f"🔄 Merging {branch_name}...")
    
    # Try to merge
    success, output, error = run_command(f"git merge origin/{branch_name} --no-edit -m 'Merge {branch_name}'")
    
    if success:
        print(f"✅ Successfully merged {branch_name}")
        return True
    else:
        # Check if it's already up to date
        if "Already up to date" in error:
            print(f"ℹ️  {branch_name} already up to date")
            return True
        elif "CONFLICT" in error or "conflict" in error:
            print(f"⚠️  Conflict in {branch_name}, resolving...")
            return resolve_conflicts(branch_name)
        else:
            print(f"❌ Failed to merge {branch_name}: {error}")
            return False

def resolve_conflicts(branch_name):
    """Resolve merge conflicts by accepting main branch version"""
    # Accept our version (main branch)
    success, output, error = run_command("git checkout --ours .")
    if not success:
        print(f"❌ Failed to resolve conflicts for {branch_name}")
        return False
    
    # Add resolved files
    success, output, error = run_command("git add .")
    if not success:
        print(f"❌ Failed to add resolved files for {branch_name}")
        return False
    
    # Commit the merge
    success, output, error = run_command(f"git commit --no-edit -m 'Merge {branch_name} (conflicts resolved)'")
    if success:
        print(f"✅ Resolved conflicts and merged {branch_name}")
        return True
    else:
        print(f"❌ Failed to commit resolved conflicts for {branch_name}")
        return False

def main():
    """Main execution function"""
    print(f"🚀 Starting efficient branch merge process at {datetime.now()}")
    print("=" * 80)
    
    # Ensure we're on main branch
    print("📍 Ensuring we're on main branch...")
    success, output, error = run_command("git checkout main")
    if not success:
        print(f"❌ Failed to checkout main: {error}")
        return 1
    
    # Pull latest changes
    print("📥 Pulling latest changes...")
    success, output, error = run_command("git pull origin main")
    if not success:
        print(f"⚠️  Warning: Failed to pull latest changes: {error}")
    
    # Get available branches
    print("📋 Getting available branches...")
    branches = get_available_branches()
    print(f"Found {len(branches)} branches to process")
    
    if not branches:
        print("❌ No branches found to merge")
        return 1
    
    # Track results
    successful = 0
    failed = 0
    already_merged = 0
    
    # Process branches in batches
    batch_size = 50
    total_batches = (len(branches) + batch_size - 1) // batch_size
    
    for batch_num in range(total_batches):
        start_idx = batch_num * batch_size
        end_idx = min(start_idx + batch_size, len(branches))
        batch_branches = branches[start_idx:end_idx]
        
        print(f"\n📦 Processing batch {batch_num + 1}/{total_batches} ({len(batch_branches)} branches)")
        print("-" * 60)
        
        for i, branch in enumerate(batch_branches):
            print(f"[{start_idx + i + 1}/{len(branches)}] ", end="")
            
            if merge_branch(branch):
                if "already up to date" in str(branch).lower():
                    already_merged += 1
                else:
                    successful += 1
            else:
                failed += 1
            
            # Small delay to avoid overwhelming git
            time.sleep(0.1)
        
        # Push changes after each batch
        if successful > 0 or already_merged > 0:
            print(f"\n📤 Pushing changes after batch {batch_num + 1}...")
            push_success, push_output, push_error = run_command("git push origin main")
            if push_success:
                print("✅ Successfully pushed changes")
            else:
                print(f"⚠️  Push failed: {push_error}")
    
    # Final summary
    print(f"\n{'='*80}")
    print("📊 MERGE SUMMARY")
    print(f"{'='*80}")
    print(f"Total branches processed: {len(branches)}")
    print(f"✅ Successfully merged: {successful}")
    print(f"ℹ️  Already up to date: {already_merged}")
    print(f"❌ Failed to merge: {failed}")
    
    # Final push
    if successful > 0 or already_merged > 0:
        print(f"\n📤 Final push to origin/main...")
        push_success, push_output, push_error = run_command("git push origin main")
        if push_success:
            print("✅ All changes pushed successfully!")
        else:
            print(f"❌ Final push failed: {push_error}")
    
    print(f"\n🏁 Merge process completed at {datetime.now()}")
    return 0 if failed == 0 else 1

if __name__ == "__main__":
    try:
        sys.exit(main())
    except KeyboardInterrupt:
        print("\n\n⚠️  Interrupted by user")
        sys.exit(1)
    except Exception as e:
        print(f"\n\n💥 Fatal error: {e}")
        sys.exit(1)

File: merge_available_branches.js
#!/usr/bin/env node

import { execSync } from 'child_process';
import fs from 'fs';

// List of available branches from the fetch output
const availableBranches = [
  'cursor/fix-errors-and-merge-to-main-0b27',
  'cursor/fix-errors-and-merge-to-main-1d7f',
  'cursor/fix-errors-and-merge-to-main-1eb2',
  'cursor/fix-errors-and-merge-to-main-41d1',
  'cursor/fix-errors-and-merge-to-main-4779',
  'cursor/fix-errors-and-merge-to-main-53b0',
  'cursor/fix-errors-and-merge-to-main-5ccc',
  'cursor/fix-errors-and-merge-to-main-687e',
  'cursor/fix-errors-and-merge-to-main-7735',
  'cursor/fix-errors-and-merge-to-main-908f',
  'cursor/fix-errors-and-merge-to-main-9203',
  'cursor/fix-errors-and-merge-to-main-aeb1',
  'cursor/fix-errors-and-merge-to-main-bc27',
  'cursor/fix-errors-and-merge-to-main-c272',
  'cursor/fix-errors-and-merge-to-main-c320',
  'cursor/fix-errors-and-merge-to-main-c358',
  'cursor/fix-errors-and-merge-to-main-c532',
  'cursor/fix-errors-and-merge-to-main-c6f4',
  'cursor/fix-errors-and-merge-to-main-c797',
  'cursor/fix-errors-and-merge-to-main-c7b1',
  'cursor/fix-errors-and-merge-to-main-ca37',
  'cursor/fix-errors-and-merge-to-main-cca8',
  'cursor/fix-errors-and-merge-to-main-d1ec',
  'cursor/fix-errors-and-merge-to-main-ee7a',
  'cursor/fix-errors-and-merge-to-main-f43c',
  'cursor/fix-errors-and-merge-to-main-f59f',
  'cursor/fix-errors-and-merge-to-main-f8eb'
];

const results = {
  successful: [],
  failed: [],
  skipped: []
};

function runCommand(command, description) {
  try {
    console.log(`\n🔄 ${description}...`);
    const output = execSync(command, { 
      encoding: 'utf8', 
      stdio: 'pipe',
      cwd: '/workspace'
    });
    console.log(`✅ ${description} completed`);
    return { success: true, output };
  } catch (error) {
    console.log(`❌ ${description} failed: ${error.message}`);
    return { success: false, error: error.message, output: error.stdout || error.stderr };
  }
}

function mergeBranch(branchName) {
  console.log(`\n📋 Processing branch: ${branchName}`);
  
  // Checkout the branch
  const checkoutResult = runCommand(`git checkout ${branchName}`, `Checkout branch ${branchName}`);
  if (!checkoutResult.success) {
    results.failed.push({ branch: branchName, reason: 'Failed to checkout branch' });
    return;
  }
  
  // Pull latest changes
  const pullResult = runCommand(`git pull origin ${branchName}`, `Pull latest changes for ${branchName}`);
  if (!pullResult.success) {
    console.log(`⚠️  Could not pull changes for ${branchName}, continuing...`);
  }
  
  // Switch back to main
  const switchToMain = runCommand(`git checkout main`, `Switch to main branch`);
  if (!switchToMain.success) {
    results.failed.push({ branch: branchName, reason: 'Failed to switch to main' });
    return;
  }
  
  // Try to merge the branch
  const mergeResult = runCommand(`git merge ${branchName} --no-ff -m "Merge branch: ${branchName}"`, `Merge ${branchName} into main`);
  
  if (mergeResult.success) {
    console.log(`✅ Successfully merged ${branchName}`);
    results.successful.push({ branch: branchName });
    
    // Push the changes
    const pushResult = runCommand(`git push origin main`, `Push merged changes to main`);
    if (!pushResult.success) {
      console.log(`⚠️  Merge successful but push failed for ${branchName}`);
    }
  } else {
    // Check if it's a conflict or other issue
    if (mergeResult.output.includes('conflict') || mergeResult.output.includes('CONFLICT')) {
      console.log(`⚠️  Merge conflict detected for ${branchName}, attempting to resolve...`);
      
      // Try to resolve conflicts automatically
      const resolveResult = runCommand(`git status --porcelain`, `Check git status for conflicts`);
      if (resolveResult.success) {
        const conflictedFiles = resolveResult.output.split('\n')
          .filter(line => line.includes('UU') || line.includes('AA') || line.includes('DD'))
          .map(line => line.split(' ').pop());
        
        console.log(`📝 Found conflicted files: ${conflictedFiles.join(', ')}`);
        
        // Try to resolve conflicts by accepting incoming changes
        for (const file of conflictedFiles) {
          if (file.trim()) {
            const resolveFile = runCommand(`git checkout --theirs "${file.trim()}"`, `Resolve conflict in ${file.trim()}`);
            if (resolveFile.success) {
              runCommand(`git add "${file.trim()}"`, `Stage resolved file ${file.trim()}`);
            }
          }
        }
        
        // Try to commit the resolved conflicts
        const commitResult = runCommand(`git commit -m "Resolve merge conflicts for ${branchName}"`, `Commit resolved conflicts`);
        if (commitResult.success) {
          console.log(`✅ Successfully resolved conflicts for ${branchName}`);
          results.successful.push({ branch: branchName });
          
          // Push the changes
          const pushResult = runCommand(`git push origin main`, `Push resolved changes to main`);
          if (!pushResult.success) {
            console.log(`⚠️  Conflict resolution successful but push failed for ${branchName}`);
          }
        } else {
          console.log(`❌ Failed to commit resolved conflicts for ${branchName}`);
          // Abort the merge
          runCommand(`git merge --abort`, `Abort failed merge for ${branchName}`);
          results.failed.push({ branch: branchName, reason: 'Failed to resolve merge conflicts' });
        }
      } else {
        results.failed.push({ branch: branchName, reason: 'Merge conflict - could not resolve' });
      }
    } else {
      console.log(`❌ Merge failed for ${branchName}: ${mergeResult.error}`);
      results.failed.push({ branch: branchName, reason: mergeResult.error });
    }
  }
}

async function main() {
  console.log('🚀 Starting comprehensive branch merge process...');
  console.log(`📊 Found ${availableBranches.length} branches to process`);
  
  // Ensure we're on main branch
  runCommand(`git checkout main`, 'Ensure on main branch');
  runCommand(`git pull origin main`, 'Pull latest main changes');
  
  // Process each branch
  for (const branchName of availableBranches) {
    try {
      mergeBranch(branchName);
    } catch (error) {
      console.log(`❌ Unexpected error processing ${branchName}: ${error.message}`);
      results.failed.push({ branch: branchName, reason: `Unexpected error: ${error.message}` });
    }
  }
  
  // Generate summary report
  console.log('\n📊 MERGE SUMMARY REPORT');
  console.log(`✅ Successful merges: ${results.successful.length}`);
  console.log(`❌ Failed merges: ${results.failed.length}`);
  console.log(`⏭️  Skipped: ${results.skipped.length}`);
  
  if (results.successful.length > 0) {
    console.log('\n✅ Successfully merged branches:');
    results.successful.forEach(({ branch }) => console.log(`  - ${branch}`));
  }
  
  if (results.failed.length > 0) {
    console.log('\n❌ Failed to merge branches:');
    results.failed.forEach(({ branch, reason }) => console.log(`  - ${branch}: ${reason}`));
  }
  
  // Save detailed results
  const reportData = {
    timestamp: new Date().toISOString(),
    summary: {
      total: availableBranches.length,
      successful: results.successful.length,
      failed: results.failed.length,
      skipped: results.skipped.length
    },
    successful: results.successful,
    failed: results.failed,
    skipped: results.skipped
  };
  
  fs.writeFileSync('/workspace/merge_available_branches_results.json', JSON.stringify(reportData, null, 2));
  console.log('\n💾 Detailed results saved to merge_available_branches_results.json');
  
  console.log('\n🎉 Branch merge process completed!');
}

main().catch(console.error);